src/lib.rs

```diff
@@ -398,9 +400,32 @@ pub async fn call<'a, T: hyper::client::connect::Connect + Clone + Send + Sync +
         client_ip
     );
 
+    let request_upgraded = request.extensions_mut().remove::<OnUpgrade>();
+
     let proxied_request = create_proxied_request(client_ip, forward_uri, request)?;
+    let mut response = client.request(proxied_request).await?;
+
+    if response.status() == StatusCode::SWITCHING_PROTOCOLS {
+        let mut response_upgraded = response
+            .extensions_mut()
+            .remove::<OnUpgrade>()
+            .expect("response does not have an upgrade extension")
+            .await?;
+
+        tokio::spawn(async move {
+            let mut request_upgraded = request_upgraded
+                .expect("request does not have an upgrade extension")
```

Comment:
I tried your code and everything works mostly well. However, with some malformed requests I managed to crash several worker threads. Perhaps there is a better way of handling errors here instead of crashing the worker thread straight away?

Reply:
I added more tests for these. The only case that is not handled now is getting an upgrade on the response, but I'm having a hard time creating a test where the server upgrades the response with being requested to do so